src/bsp/src/stm32/src/Decawave.cpp

```diff
@@ -0,0 +1,307 @@
+    configureDW();
+}
+
+void Decawave::setLed(DW_LED led, bool enabled) {
+    uint8_t dwGPIO = 0;
+    uint8_t dwGPIOValue = 0;
+
+    switch (led) {
+    case DW_LED::LED_0:
+        dwGPIO = DWT_GxM0;
+        if (enabled) {
+            dwGPIOValue = DWT_GxP0;
+        }
+        break;
+
+    case DW_LED::LED_1:
+        dwGPIO = DWT_GxM1;
+        if (enabled) {
+            dwGPIOValue = DWT_GxP1;
+        }
+        break;
+
+    case DW_LED::LED_2:
+        dwGPIO = DWT_GxM2;
+        if (enabled) {
+            dwGPIOValue = DWT_GxP2;
+        }
+        break;
+
+    case DW_LED::LED_3:
+        dwGPIO = DWT_GxM3;
+        if (enabled) {
+            dwGPIOValue = DWT_GxP3;
+        }
+        break;
+    }
+
+    deca_selectDevice(m_spiDevice);
+    dwt_setgpiovalue(dwGPIO, dwGPIOValue);
+}
+
+void Decawave::receiveInternal(UWBRxFrame& frame,
+                               uint16_t timeoutUs,
+                               uint8_t flags,
+                               bool rxStarted) {
+    m_rxTaskHandle = xTaskGetCurrentTaskHandle();
+    deca_selectDevice(m_spiDevice);
+
+    frame.m_status = UWBRxStatus::ONGOING;
+
+    if (!rxStarted) {
+        dwt_setrxtimeout(timeoutUs);
+        dwt_rxenable(flags);
+    }
+
+    ulTaskNotifyTake(pdTRUE, portMAX_DELAY);
+
+    retrieveRxFrame(&frame);
+}
+
+void Decawave::receiveAsyncInternal(UWBRxFrame& frame,
+                                    uint16_t timeoutUs,
+                                    uint8_t flags,
+                                    bool rxStarted) {
+    m_rxFrame = &frame;
+    m_rxTaskHandle = m_rxAsyncTask.getTaskHandle();
+
+    deca_selectDevice(m_spiDevice);
+
+    frame.m_status = UWBRxStatus::ONGOING;
+
+    if (!rxStarted) {
+        dwt_setrxtimeout(timeoutUs);
+        dwt_rxenable(flags);
+    }
+}
+
+void Decawave::receive(UWBRxFrame& frame, uint16_t timeoutUs) {
+    receiveInternal(frame, timeoutUs, DWT_START_TX_IMMEDIATE);
+}
+
+void Decawave::receiveDelayed(UWBRxFrame& frame, uint16_t timeoutUs, uint64_t rxStartTime) {
+    deca_selectDevice(m_spiDevice);
+    dwt_setdelayedtrxtime(rxStartTime >> 8);
+
+    receiveInternal(frame, timeoutUs, rxStartTime);
+}
+
+void Decawave::receiveAsync(UWBRxFrame& frame, uint16_t timeoutUs) {
+    receiveAsyncInternal(frame, timeoutUs, DWT_START_TX_IMMEDIATE);
+}
+
+void Decawave::receiveAsyncDelayed(UWBRxFrame& frame, uint16_t timeoutUs, uint64_t rxStartTime) {
+    deca_selectDevice(m_spiDevice);
+    dwt_setdelayedtrxtime(rxStartTime >> 8);
+
+    receiveAsyncInternal(frame, timeoutUs, rxStartTime);
+}
+
+bool Decawave::transmitInternal(uint8_t* buf, uint16_t length, uint8_t flags) {
+    if (length > UWB_MAX_LENGTH) {
+        return false;
+    }
+    deca_selectDevice(m_spiDevice);
+
+    memcpy(m_txBuffer.data(), buf, length);
+
+    // Send two bytes more than requested because of the DW auto-generated CRC16
+    dwt_writetxdata(length + 2, m_txBuffer.data(), 0);
+    dwt_writetxfctrl(length + 2, 0, 0);
```

Comment:
Could you put the 2 in a define or constexpr, would make the code more auto documented